# Libraries and packages

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib as mpl
import random
import time
#PATH PROCESS
import os
import os.path
from pathlib import Path
import glob
from scipy.io import loadmat
import nibabel as nib
import csv
#IMAGE PROCESS
from PIL import Image
from keras.preprocessing import image
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import cv2
from keras.applications.vgg16 import preprocess_input, decode_predictions
from keras.preprocessing import image
from skimage.feature import hessian_matrix, hessian_matrix_eigvals
from scipy.ndimage import convolve
from skimage import data, io, filters
import skimage
from skimage.morphology import convex_hull_image, erosion
from IPython import display
from scipy.ndimage import gaussian_filter
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
import matplotlib.patches as patches
#SCALER & TRANSFORMATION
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from keras import regularizers
from sklearn.preprocessing import LabelEncoder
#ACCURACY CONTROL
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score, roc_curve
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.metrics import mean_squared_error, r2_score
#OPTIMIZER
from keras.optimizers import RMSprop,Adam,Optimizer,Optimizer, SGD
#MODEL LAYERS
from tensorflow.keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPool2D, BatchNormalization,MaxPooling2D,BatchNormalization,\
                        Permute, TimeDistributed, Bidirectional,GRU, SimpleRNN,\
LSTM, GlobalAveragePooling2D, SeparableConv2D, ZeroPadding2D, Convolution2D, ZeroPadding2D,Reshape, Conv2DTranspose,\
LeakyReLU, GaussianNoise, GlobalMaxPooling2D, ReLU, Input, Concatenate
from keras import models
from keras import layers
import tensorflow as tf
from keras.applications import VGG16,VGG19,inception_v3
from keras import backend as K
from keras.utils import plot_model
from keras.datasets import mnist
import keras
from keras.models import Model
#IGNORING WARNINGS
from warnings import filterwarnings
filterwarnings("ignore",category=DeprecationWarning)
filterwarnings("ignore", category=FutureWarning) 
filterwarnings("ignore", category=UserWarning)

from keras.layers import Input, Conv2D, MaxPooling2D, Conv2DTranspose, concatenate, BatchNormalization, Activation, add
from keras.models import Model, model_from_json
from keras.optimizers import Adam
from tensorflow.keras.layers import ELU, LeakyReLU
from tensorflow.keras.utils import plot_model
from tensorflow.keras import backend as K

random.seed(42)

%matplotlib inline

# Dataset preprocessing

## Training and validation set preprocessing

### Path to labels and original MRI Images for the training set

In [2]:
# Label100_Path = Path("../input/mrihs/label/100label")
# Image100_Path = Path("../input/mrihs/original/100")

### Extract all the jpg images

In [3]:
# Label_Series = pd.Series(list(Label100_Path.glob(r"**/*.jpg")),name="LABEL").astype(str)
# Image_Series = pd.Series(list(Image100_Path.glob(r"**/*.jpg")),name="IMAGE").astype(str)

In [4]:
# print("LABEL LEN:",len(Label_Series))
# print("---"*10)
# print("IMAGE LEN:",len(Image_Series))
# print("---"*10)
# print("LABEL TYPE:",type(Label_Series))
# print("---"*10)
# print("IMAGE TYPE:",type(Image_Series))
# print("---"*10)

In [5]:
# print("LABEL PATH EXAMPLE:\n",Label_Series[0])
# print("---"*10)
# print("IMAGE PATH EXAMPLE:\n",Image_Series[0])

### Splliting image filepaths

### Listing all segmentation masks

In [6]:
# L_IMG = []
# R_IMG = []

# Split_Params_For_LABEL = "100label/"
# Common_Params = "/"
# List_Split_Params = "_"

# for label_x in Label_Series:
#     try:
        
#         L_Main_Path,L_Target_Path = label_x.split(Split_Params_For_LABEL)

#         Label_Path_Before,Label_Path_Middle,Label_Path_After = L_Target_Path.split(Common_Params)

#         if Label_Path_Before == 'ADNI_013_S_0325_85153_ACPC':
#             continue # right label is missing
        
#         Label_Path_Split = Label_Path_After.split(List_Split_Params)

#         if Label_Path_Split[-2] == "L":
            
#             L_IMG.append(label_x)
            
#         elif Label_Path_Split[-2] == "R":
            
#             R_IMG.append(label_x)
            
#         else:
#             print(f"SOMETHING IS WRONG for {label_x}")
            
#     except:
#         print("LABEL: ",label_x)

In [7]:
# print("LABEL L:",len(L_IMG))
# print("LABEL R:",len(R_IMG))

### Listing all MRI images

In [8]:
# Total_IMG = []
# Split_Params_For_IMG = "100/"

# for image_x in Image_Series:
#     try:
        
#         I_Main_Path,I_Target_Path = image_x.split(Split_Params_For_IMG)

#         IMG_Path_Before,IMG_Path_After = I_Target_Path.split(Common_Params)

#         if IMG_Path_Before == 'ADNI_013_S_0325_85153_ACPC':
#             continue # right label is missing

#         IMG_Path_Split = IMG_Path_After.split(List_Split_Params)

#         Total_IMG.append(image_x)
            
#     except:
#         print("IMG: ",image_x)

In [9]:
# print("IMG :",len(Total_IMG))

In [10]:
# L_IMG = sorted(L_IMG)
# R_IMG = sorted(R_IMG)
# Total_IMG = sorted(Total_IMG)

### Getting all images
### Merge left and right segmentation masks into one mask
### Resize all images to (128,128)

In [11]:
# X = []
# Y = []

# for i in range(len(Total_IMG)):
#     x_l = L_IMG[i]
#     x_r = R_IMG[i]
#     x_image = Total_IMG[i] 
    
#     img = cv2.imread(x_image, cv2.IMREAD_COLOR)
#     resized_img = cv2.resize(img,(128, 128), interpolation = cv2.INTER_CUBIC)
    
#     X.append(resized_img)
    
#     msk_l = cv2.imread(x_l, cv2.IMREAD_GRAYSCALE)
#     msk_r = cv2.imread(x_r, cv2.IMREAD_GRAYSCALE)
#     combined_mask = msk_l + msk_r
#     combined_mask = np.clip(combined_mask, 0, 255).astype(np.uint8)
    
#     resized_msk = cv2.resize(combined_mask,(128, 128), interpolation = cv2.INTER_CUBIC)
    
#     Y.append(resized_msk)

In [12]:
# print(len(X))
# print(len(Y))

### Convert images into numpy arrays for downstream analyses

In [13]:
# X = np.array(X, dtype='float32')
# Y = np.array(Y, dtype='float32')

In [14]:
# print(X.shape)
# print(Y.shape)

### Save preprocessed dataset

In [15]:
# np.savez_compressed('X_Y_compressed_128_128.npz', X=X, Y=Y)
# print("Saved compressed X and Y successfully")

### Load preprocessed training set images

In [16]:
# Load compressed file
data = np.load('X_Y_compressed_128_128.npz')

# Access arrays
X = data['X']
Y = data['Y']

print("X shape:", X.shape)
print("Y shape:", Y.shape)

X shape: (18711, 128, 128, 3)
Y shape: (18711, 128, 128)


### Prepare training and validation set

In [17]:
X_train, X_val, Y_train, Y_val = train_test_split(X, Y, test_size=0.2, random_state=42)

### Free unused variables

In [18]:
del X,Y

### Normalize the images to [0,1] range

In [19]:
Y_train = Y_train.reshape((Y_train.shape[0],Y_train.shape[1],Y_train.shape[2],1))
Y_val = Y_val.reshape((Y_val.shape[0],Y_val.shape[1],Y_val.shape[2],1))

X_train = X_train / 255
X_val = X_val / 255
Y_train = Y_train / 255
Y_val = Y_val / 255

Y_train = np.round(Y_train,0)	
Y_val = np.round(Y_val,0)	

print("X_train.shape:", X_train.shape)
print("X_val.shape:", X_val.shape)
print("Y_train.shape:", Y_train.shape)
print("Y_val.shape:", Y_val.shape)

print("Max value in X_train:", np.max(X_train))
print("Min value in X_train:", np.min(X_train))
print("Max value in X_val:", np.max(X_val))
print("Min value in X_val:", np.min(X_val))
print("Max value in Y_train:", np.max(Y_train))
print("Min value in Y_train:", np.min(Y_train))
print("Max value in Y_val:", np.max(Y_val))
print("Min value in Y_val:", np.min(Y_val))

X_train.shape: (14968, 128, 128, 3)
X_val.shape: (3743, 128, 128, 3)
Y_train.shape: (14968, 128, 128, 1)
Y_val.shape: (3743, 128, 128, 1)
Max value in X_train: 1.0
Min value in X_train: 0.0
Max value in X_val: 1.0
Min value in X_val: 0.0
Max value in Y_train: 1.0
Min value in Y_train: 0.0
Max value in Y_val: 1.0
Min value in Y_val: 0.0


## Test set

### Preprocess test set images

In [20]:
# Label35_Path = Path("../input/mrihs/label/35label")
# Image35_Path = Path("../input/mrihs/original/35")

# Label_Series = pd.Series(list(Label35_Path.glob(r"**/*.jpg")),name="LABEL").astype(str)
# Image_Series = pd.Series(list(Image35_Path.glob(r"**/*.jpg")),name="IMAGE").astype(str)

# print("LABEL LEN:",len(Label_Series))
# print("---"*10)
# print("IMAGE LEN:",len(Image_Series))
# print("---"*10)

# L_IMG = []
# R_IMG = []

# Split_Params_For_LABEL = "35label/"
# Common_Params = "/"
# List_Split_Params = "_"

# for label_x in Label_Series:
#     try:
#         L_Main_Path,L_Target_Path = label_x.split(Split_Params_For_LABEL)
#         Label_Path_Before,Label_Path_Middle,Label_Path_After = L_Target_Path.split(Common_Params)
#         Label_Path_Split = Label_Path_After.split(List_Split_Params)
        
#         if Label_Path_Split[-2] == "L":    
#             L_IMG.append(label_x)
#         elif Label_Path_Split[-2] == "R":
#             R_IMG.append(label_x)
#         else:
#             print(f"SOMETHING IS WRONG for {label_x}")
#     except:
#         print("LABEL: ",label_x)

# Total_IMG = []
# Split_Params_For_IMG = "35/"

# for image_x in Image_Series:
#     try:
#         I_Main_Path,I_Target_Path = image_x.split(Split_Params_For_IMG)
#         IMG_Path_Before,IMG_Path_After = I_Target_Path.split(Common_Params)
#         IMG_Path_Split = IMG_Path_After.split(List_Split_Params)
#         Total_IMG.append(image_x)
#     except:
#         print("IMG: ",image_x)

# L_IMG = sorted(L_IMG)
# R_IMG = sorted(R_IMG)
# Total_IMG = sorted(Total_IMG)

# print("LABEL L:",len(L_IMG))
# print("LABEL R:",len(R_IMG))
# print("IMG :",len(Total_IMG))

# X_test = []
# Y_test = []

# for i in range(len(Total_IMG)):
#     x_l = L_IMG[i]
#     x_r = R_IMG[i]
#     x_image = Total_IMG[i] 
    
#     img = cv2.imread(x_image, cv2.IMREAD_COLOR)
#     resized_img = cv2.resize(img,(128, 128), interpolation = cv2.INTER_CUBIC)
    
#     X_test.append(resized_img)
    
#     msk_l = cv2.imread(x_l, cv2.IMREAD_GRAYSCALE)
#     msk_r = cv2.imread(x_r, cv2.IMREAD_GRAYSCALE)
#     combined_mask = msk_l + msk_r
#     combined_mask = np.clip(combined_mask, 0, 255).astype(np.uint8)
    
#     resized_msk = cv2.resize(combined_mask,(128, 128), interpolation = cv2.INTER_CUBIC)
    
#     Y_test.append(resized_msk)

# print('Length of X_test:', len(X_test))
# print('Length of Y_test:', len(Y_test))

# X_test = np.array(X_test, dtype='float32')
# Y_test = np.array(Y_test, dtype='float32')

# print('Shape of X_test:', X_test.shape)
# print('Shape of Y_test:', Y_test.shape)

# np.savez_compressed('X_Y_test_compressed_128_128.npz', X_test=X_test, Y_test=Y_test)
# print("Saved compressed X and Y test set successfully")

### Load preprocessed test set

In [21]:
# Load compressed file
data = np.load('X_Y_test_compressed_128_128.npz')

# Access arrays
X_test = data['X_test']
Y_test = data['Y_test']

print("X_test shape:", X_test.shape)
print("Y_test shape:", Y_test.shape)

X_test shape: (6615, 128, 128, 3)
Y_test shape: (6615, 128, 128)


In [22]:
print("Max, min value in X_test:", np.max(X_test), np.min(X_test))
print("Max, min value in Y_test:", np.max(Y_test), np.min(Y_test))

Max, min value in X_test: 255.0 0.0
Max, min value in Y_test: 255.0 0.0


### Reshape and normalize the test set

In [23]:
Y_test = Y_test.reshape((Y_test.shape[0],Y_test.shape[1],Y_test.shape[2],1))

X_test = X_test / 255
Y_test = Y_test / 255

Y_test = np.round(Y_test,0)	

print("X_test.shape:", X_test.shape)
print("Y_test.shape:", Y_test.shape)

print("Max value in X_test:", np.max(X_test))
print("Min value in X_test:", np.min(X_test))
print("Max value in Y_test:", np.max(Y_test))
print("Min value in Y_test:", np.min(Y_test))

X_test.shape: (6615, 128, 128, 3)
Y_test.shape: (6615, 128, 128, 1)
Max value in X_test: 1.0
Min value in X_test: 0.0
Max value in Y_test: 1.0
Min value in Y_test: 0.0


# Models

## MultiResUNet

- [Paper](https://doi.org/10.1016/j.neunet.2019.08.025)
- [Codebase](https://github.com/nibtehaz/MultiResUNet/tree/master)

In [24]:
def conv2d_bn(x, filters, num_row, num_col, padding='same', strides=(1, 1), activation='relu', name=None):
    '''
    2D Convolutional layers
    
    Arguments:
        x {keras layer} -- input layer 
        filters {int} -- number of filters
        num_row {int} -- number of rows in filters
        num_col {int} -- number of columns in filters
    
    Keyword Arguments:
        padding {str} -- mode of padding (default: {'same'})
        strides {tuple} -- stride of convolution operation (default: {(1, 1)})
        activation {str} -- activation function (default: {'relu'})
        name {str} -- name of the layer (default: {None})
    
    Returns:
        [keras layer] -- [output layer]
    '''

    x = Conv2D(filters, (num_row, num_col), strides=strides, padding=padding, use_bias=False)(x)
    x = BatchNormalization(axis=3, scale=False)(x)

    if(activation == None):
        return x

    x = Activation(activation, name=name)(x)

    return x


def trans_conv2d_bn(x, filters, num_row, num_col, padding='same', strides=(2, 2), name=None):
    '''
    2D Transposed Convolutional layers
    
    Arguments:
        x {keras layer} -- input layer 
        filters {int} -- number of filters
        num_row {int} -- number of rows in filters
        num_col {int} -- number of columns in filters
    
    Keyword Arguments:
        padding {str} -- mode of padding (default: {'same'})
        strides {tuple} -- stride of convolution operation (default: {(2, 2)})
        name {str} -- name of the layer (default: {None})
    
    Returns:
        [keras layer] -- [output layer]
    '''

    x = Conv2DTranspose(filters, (num_row, num_col), strides=strides, padding=padding)(x)
    x = BatchNormalization(axis=3, scale=False)(x)
    
    return x


def MultiResBlock(U, inp, alpha = 1.67):
    '''
    MultiRes Block
    
    Arguments:
        U {int} -- Number of filters in a corrsponding UNet stage
        inp {keras layer} -- input layer 
    
    Returns:
        [keras layer] -- [output layer]
    '''

    W = alpha * U

    shortcut = inp

    shortcut = conv2d_bn(shortcut, int(W*0.167) + int(W*0.333) +
                         int(W*0.5), 1, 1, activation=None, padding='same')

    conv3x3 = conv2d_bn(inp, int(W*0.167), 3, 3,
                        activation='relu', padding='same')

    conv5x5 = conv2d_bn(conv3x3, int(W*0.333), 3, 3,
                        activation='relu', padding='same')

    conv7x7 = conv2d_bn(conv5x5, int(W*0.5), 3, 3,
                        activation='relu', padding='same')

    out = concatenate([conv3x3, conv5x5, conv7x7], axis=3)
    out = BatchNormalization(axis=3)(out)

    out = add([shortcut, out])
    out = Activation('relu')(out)
    out = BatchNormalization(axis=3)(out)

    return out


def ResPath(filters, length, inp):
    '''
    ResPath
    
    Arguments:
        filters {int} -- [description]
        length {int} -- length of ResPath
        inp {keras layer} -- input layer 
    
    Returns:
        [keras layer] -- [output layer]
    '''


    shortcut = inp
    shortcut = conv2d_bn(shortcut, filters, 1, 1,
                         activation=None, padding='same')

    out = conv2d_bn(inp, filters, 3, 3, activation='relu', padding='same')

    out = add([shortcut, out])
    out = Activation('relu')(out)
    out = BatchNormalization(axis=3)(out)

    for i in range(length-1):

        shortcut = out
        shortcut = conv2d_bn(shortcut, filters, 1, 1,
                             activation=None, padding='same')

        out = conv2d_bn(out, filters, 3, 3, activation='relu', padding='same')

        out = add([shortcut, out])
        out = Activation('relu')(out)
        out = BatchNormalization(axis=3)(out)

    return out


def MultiResUnet(height, width, n_channels):
    '''
    MultiResUNet
    
    Arguments:
        height {int} -- height of image 
        width {int} -- width of image 
        n_channels {int} -- number of channels in image
    
    Returns:
        [keras model] -- MultiResUNet model
    '''


    inputs = Input((height, width, n_channels))

    mresblock1 = MultiResBlock(32, inputs)
    pool1 = MaxPooling2D(pool_size=(2, 2))(mresblock1)
    mresblock1 = ResPath(32, 4, mresblock1)

    mresblock2 = MultiResBlock(32*2, pool1)
    pool2 = MaxPooling2D(pool_size=(2, 2))(mresblock2)
    mresblock2 = ResPath(32*2, 3, mresblock2)

    mresblock3 = MultiResBlock(32*4, pool2)
    pool3 = MaxPooling2D(pool_size=(2, 2))(mresblock3)
    mresblock3 = ResPath(32*4, 2, mresblock3)

    mresblock4 = MultiResBlock(32*8, pool3)
    pool4 = MaxPooling2D(pool_size=(2, 2))(mresblock4)
    mresblock4 = ResPath(32*8, 1, mresblock4)

    mresblock5 = MultiResBlock(32*16, pool4)

    up6 = concatenate([Conv2DTranspose(
        32*8, (2, 2), strides=(2, 2), padding='same')(mresblock5), mresblock4], axis=3)
    mresblock6 = MultiResBlock(32*8, up6)

    up7 = concatenate([Conv2DTranspose(
        32*4, (2, 2), strides=(2, 2), padding='same')(mresblock6), mresblock3], axis=3)
    mresblock7 = MultiResBlock(32*4, up7)

    up8 = concatenate([Conv2DTranspose(
        32*2, (2, 2), strides=(2, 2), padding='same')(mresblock7), mresblock2], axis=3)
    mresblock8 = MultiResBlock(32*2, up8)

    up9 = concatenate([Conv2DTranspose(32, (2, 2), strides=(
        2, 2), padding='same')(mresblock8), mresblock1], axis=3)
    mresblock9 = MultiResBlock(32, up9)

    conv10 = conv2d_bn(mresblock9, 1, 1, 1, activation='sigmoid')
    
    model = Model(inputs=[inputs], outputs=[conv10])

    return model

## Modified U-Net

- [Paper](https://pubmed.ncbi.nlm.nih.gov/35304675/)

In [25]:
def conv_block(input_tensor, filters, kernel_sizes=[1, 3, 5]):
    """Convolutional block with filters of multiple kernel sizes."""
    convs = []
    for k in kernel_sizes:
        conv = layers.Conv2D(filters, (k, k), padding='same', activation='relu')(input_tensor)
        convs.append(conv)
    merged = layers.Concatenate()(convs)
    bn = layers.BatchNormalization()(merged)
    return bn

def modified_unet(input_shape=(128, 128, 3)):
    """Builds the modified U-Net model as described in the paper."""
    inputs = layers.Input(shape=input_shape)

    # Contracting Path
    c1 = conv_block(inputs, 64)
    p1 = layers.MaxPooling2D((2, 2))(c1)
    p1 = layers.Dropout(0.1)(p1)

    c2 = conv_block(p1, 128)
    p2 = layers.MaxPooling2D((2, 2))(c2)
    p2 = layers.Dropout(0.1)(p2)

    c3 = conv_block(p2, 256)
    p3 = layers.MaxPooling2D((2, 2))(c3)
    p3 = layers.Dropout(0.1)(p3)

    c4 = conv_block(p3, 512)
    p4 = layers.MaxPooling2D((2, 2))(c4)
    p4 = layers.Dropout(0.1)(p4)

    # Bottleneck
    bn = conv_block(p4, 1024)

    # Expansive Path
    u4 = layers.Conv2DTranspose(512, (2, 2), strides=(2, 2), padding='same')(bn)
    u4 = layers.Concatenate()([u4, c4])
    u4 = layers.Dropout(0.1)(u4)
    c5 = conv_block(u4, 512)

    u3 = layers.Conv2DTranspose(256, (2, 2), strides=(2, 2), padding='same')(c5)
    u3 = layers.Concatenate()([u3, c3])
    u3 = layers.Dropout(0.1)(u3)
    c6 = conv_block(u3, 256)

    u2 = layers.Conv2DTranspose(128, (2, 2), strides=(2, 2), padding='same')(c6)
    u2 = layers.Concatenate()([u2, c2])
    u2 = layers.Dropout(0.1)(u2)
    c7 = conv_block(u2, 128)

    u1 = layers.Conv2DTranspose(64, (2, 2), strides=(2, 2), padding='same')(c7)
    u1 = layers.Concatenate()([u1, c1])
    u1 = layers.Dropout(0.1)(u1)
    c8 = conv_block(u1, 64)

    # Output Layer
    outputs = layers.Conv2D(1, (1, 1), activation='sigmoid')(c8)

    model = models.Model(inputs, outputs)
    return model

## Our Model

Key improvements:
- Swish Activation Function 
- SpatialDropout2D
- Bilinear Interpolation
- Combined Loss Function: Binary Cross-Entropy + Dice Loss

In [26]:
def our_conv_block(input_tensor, filters, kernel_sizes=[1, 3, 5], activation='swish', dropout_rate=0.1):
    """Convolutional block with filters of multiple kernel sizes and improved activation/regularization."""
    convs = []
    for k in kernel_sizes:
        conv = layers.Conv2D(filters, (k, k), padding='same', activation=activation)(input_tensor)
        conv = layers.BatchNormalization()(conv)
        convs.append(conv)
    merged = layers.Concatenate()(convs)
    merged = layers.SpatialDropout2D(dropout_rate)(merged)
    return merged

def our_modified_unet(input_shape=(128, 128, 3)):
    """Builds the modified U-Net model with updated activation, regularization, upsampling, and loss functions."""
    inputs = layers.Input(shape=input_shape)

    # Contracting Path
    c1 = our_conv_block(inputs, 64)
    p1 = layers.MaxPooling2D((2, 2))(c1)

    c2 = our_conv_block(p1, 128)
    p2 = layers.MaxPooling2D((2, 2))(c2)

    c3 = our_conv_block(p2, 256)
    p3 = layers.MaxPooling2D((2, 2))(c3)

    c4 = our_conv_block(p3, 512)
    p4 = layers.MaxPooling2D((2, 2))(c4)

    # Bottleneck
    bn = our_conv_block(p4, 1024)

    # Expansive Path
    u4 = layers.UpSampling2D((2, 2), interpolation='bilinear')(bn)
    u4 = layers.Concatenate()([u4, c4])
    c5 = our_conv_block(u4, 512)

    u3 = layers.UpSampling2D((2, 2), interpolation='bilinear')(c5)
    u3 = layers.Concatenate()([u3, c3])
    c6 = our_conv_block(u3, 256)

    u2 = layers.UpSampling2D((2, 2), interpolation='bilinear')(c6)
    u2 = layers.Concatenate()([u2, c2])
    c7 = our_conv_block(u2, 128)

    u1 = layers.UpSampling2D((2, 2), interpolation='bilinear')(c7)
    u1 = layers.Concatenate()([u1, c1])
    c8 = our_conv_block(u1, 64)

    # Output Layer
    outputs = layers.Conv2D(1, (1, 1), activation='sigmoid')(c8)

    model = models.Model(inputs, outputs)
    return model

def dice_loss(y_true, y_pred, smooth=1e-6):
    """Dice loss for binary segmentation."""
    y_true_f = tf.keras.backend.flatten(y_true)
    y_pred_f = tf.keras.backend.flatten(y_pred)
    intersection = tf.keras.backend.sum(y_true_f * y_pred_f)
    return 1 - (2. * intersection + smooth) / (tf.keras.backend.sum(y_true_f) + tf.keras.backend.sum(y_pred_f) + smooth)

def combined_loss(y_true, y_pred):
    """Combines Binary Cross-Entropy with Dice Loss."""
    bce = tf.keras.losses.binary_crossentropy(y_true, y_pred)
    d_loss = dice_loss(y_true, y_pred)
    return bce + d_loss


# Model training

In [27]:
def dice_coef(y_true, y_pred):
    smooth = 1e-6
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred)
    intersection = K.sum(y_true_f * y_pred_f)
    return (2. * intersection + smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)

def jacard(y_true, y_pred):
    smooth = 1e-6
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred)
    intersection = K.sum ( y_true_f * y_pred_f)
    union = K.sum ( y_true_f + y_pred_f - y_true_f * y_pred_f)

    return (intersection + smooth) / (union + smooth)

def saveModel(model, outfilepath='models'):

    model_json = model.to_json()

    try:
        os.makedirs(outfilepath)
    except:
        pass
    
    fp = open(f'{outfilepath}/modelP.json','w')
    fp.write(model_json)
    model.save_weights(f'{outfilepath}/modelW.weights.h5')

def evaluateModel(model, X_test, Y_test, batchSize, outfilepath='models', resfilepath='results', val=True):
    
    try:
        os.makedirs(resfilepath)
    except:
        pass 
    

    yp = model.predict(x=X_test, batch_size=batchSize, verbose=1)
    yp = np.round(yp,0)

    smooth = 1e-6
    
    jacard = 0
    dice = 0
    jacard_scores = []
    
    for i in range(len(Y_test)):
        yp_2 = yp[i].ravel()
        y2 = Y_test[i].ravel()
        
        intersection = yp_2 * y2
        union = yp_2 + y2 - intersection

        jacard_score = ((np.sum(intersection) + smooth) / (np.sum(union) + smooth))  
        jacard += jacard_score
        
        dice += (2. * np.sum(intersection) + smooth) / (np.sum(yp_2) + np.sum(y2) + smooth)

        if np.sum(yp_2) == 0 and np.sum(y2) == 0:
            continue
        jacard_scores.append((i, jacard_score))

    jacard /= len(Y_test)
    dice /= len(Y_test)
    
    print('Jacard Index : '+str(jacard))
    print('Dice Coefficient : '+str(dice))
    
    # Sort the scores by Jaccard index
    jacard_scores = sorted(jacard_scores, key=lambda x: x[1], reverse=True)
    top_bottom_10_scores = [item[0] for item in jacard_scores[:5]] + [item[0] for item in jacard_scores[-5:]]

    for idx in top_bottom_10_scores:
        # Prepare plots
        fig, axes = plt.subplots(1, 3, figsize=(12, 5))  # 3 columns: Input, Ground Truth, Prediction
        fig.suptitle("Testing Images: Input vs Ground Truth vs Prediction", fontsize=16)

        input_image = X_test[idx]  # Shape: (height, width, channels)
        ground_truth = Y_test[idx]  # Shape: (height, width, 1)
        prediction = yp[idx]
        
        # Plot input image
        axes[0].imshow(input_image.squeeze(), cmap="gray")
        axes[0].set_title("Input Image")
        axes[0].axis("off")

        # Plot ground truth
        axes[1].imshow(ground_truth.squeeze(), cmap="gray")
        axes[1].set_title("Ground Truth")
        axes[1].axis("off")

        # Plot prediction
        axes[2].imshow(prediction.squeeze(), cmap="gray")
        axes[2].set_title("Prediction")
        axes[2].axis("off")

        intersection = yp[idx].ravel() * Y_test[idx].ravel()
        union = yp[idx].ravel() + Y_test[idx].ravel() - intersection
        
        jacard_score = ((np.sum(intersection) + smooth) / (np.sum(union) + smooth))  
        plt.suptitle('Jacard Index: '+ str(np.sum(intersection)) +' / '+ str(np.sum(union)) +' = '+str(jacard_score))

        # Adjust layout
        # plt.tight_layout(rect=[0, 0, 1, 0.96])  # Leave space for the title

        # Adjust spacing to remove extra white space
        plt.subplots_adjust(wspace=0.5, hspace=0)  # Minimal horizontal and vertical spacing
        plt.tight_layout(pad=0.5)  # Remove padding around the figure
        
        # Save the figure without additional margins
        plt.savefig(resfilepath + '/' + str(i) +'.png',format='png', bbox_inches='tight', pad_inches=0.5)
        
        # plt.savefig()
        plt.close()


    if val:
        fp = open(f'{outfilepath}/log.txt','a')
        fp.write(str(jacard)+'\n')
        fp.close()
    
        fp = open(f'{outfilepath}/best.txt','r')
        best = fp.read()
        fp.close()
    
        if(jacard>float(best)):
            print('***********************************************')
            print('Jacard Index improved from '+str(best)+' to '+str(jacard))
            print('***********************************************')
            fp = open(f'{outfilepath}/best.txt','w')
            fp.write(str(jacard))
            fp.close()
    
            saveModel(model, outfilepath=outfilepath)


def trainStep(model, X_train, Y_train, X_test, Y_test, epochs, batchSize, outfilepath, resfilepath):

    
    for epoch in range(epochs):
        print('Epoch : {}'.format(epoch+1))
        model.fit(x=X_train, y=Y_train, batch_size=batchSize, epochs=1, verbose=1)     

        evaluateModel(model,X_test, Y_test,batchSize, outfilepath=outfilepath, resfilepath=resfilepath)

    return model 

# def combined_loss(y_true, y_pred):
#     smooth = 1e-6
#     # Dice loss
#     y_true_f = K.flatten(y_true)
#     y_pred_f = K.flatten(y_pred)
#     intersection = K.sum(y_true_f * y_pred_f)
#     dice_loss = 1 - (2. * intersection + smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)
#     # Binary cross-entropy
#     bce_loss = K.binary_crossentropy(y_true, y_pred)
#     return dice_loss + bce_loss

### Initiate MultiResUNet

In [28]:
# model = MultiResUnet(height=128, width=128, n_channels=3)

# model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[dice_coef, jacard, 'accuracy'])

# outfilepath = 'models' # output file path for the model
# resfilepath = 'results' # results file path for the model

### Initiate Modified U-Net

In [29]:
# model = modified_unet(input_shape=(128, 128, 3))

# # model.compile(optimizer='adam', loss=combined_loss, metrics=[dice_coef, jacard, 'accuracy'])
# model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[dice_coef, jacard, 'accuracy'])

# outfilepath = 'models_hazarika' # output file path for the model
# resfilepath = 'results_hazarika' # results file path for the model

### Initiate our model

In [30]:
model = our_modified_unet(input_shape=(128, 128, 3))

model.compile(optimizer='adam', loss=combined_loss, metrics=[dice_coef, jacard, 'accuracy'])

outfilepath = 'models_our' # output file path for the model
resfilepath = 'results_our' # results file path for the model

In [31]:
model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 128, 128,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d (Conv2D)     │ (None, 128, 128,  │        256 │ input_layer[0][0] │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_1 (Conv2D)   │ (None, 128, 128,  │      1,792 │ input_layer[0][0] │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_2 (Conv2D)   │ (None, 128, 128,  │      4,864 │ input_layer[0][0] │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 128, 128,  │        256 │ conv2d[0][0]      │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128, 128,  │        256 │ conv2d_1[0][0]    │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128, 128,  │        256 │ conv2d_2[0][0]    │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate         │ (None, 128, 128,  │          0 │ batch_normalizat… │
│ (Concatenate)       │ 192)              │            │ batch_normalizat… │
│                     │                   │            │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ spatial_dropout2d   │ (None, 128, 128,  │          0 │ concatenate[0][0] │
│ (SpatialDropout2D)  │ 192)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ max_pooling2d       │ (None, 64, 64,    │          0 │ spatial_dropout2… │
│ (MaxPooling2D)      │ 192)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_3 (Conv2D)   │ (None, 64, 64,    │     24,704 │ max_pooling2d[0]… │
│                     │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_4 (Conv2D)   │ (None, 64, 64,    │    221,312 │ max_pooling2d[0]… │
│                     │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_5 (Conv2D)   │ (None, 64, 64,    │    614,528 │ max_pooling2d[0]… │
│                     │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 64, 64,    │        512 │ conv2d_3[0][0]    │
│ (BatchNormalizatio… │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 64, 64,    │        512 │ conv2d_4[0][0]    │
│ (BatchNormalizatio… │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 64, 64,    │        512 │ conv2d_5[0][0]    │
│ (BatchNormalizatio… │ 128)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼─────────────────

 Total params: 182,835,073 (697.46 MB)

 Trainable params: 182,817,409 (697.39 MB)

 Non-trainable params: 17,664 (69.00 KB)

### Run this cell if the model is to be trained for the first time

In [32]:
# saveModel(model, outfilepath=outfilepath)

# fp = open(f'{outfilepath}/log.txt','w')
# fp.close()
# fp = open(f'{outfilepath}/best.txt','w')
# fp.write('-1.0')
# fp.close()  

In [33]:
# trainStep(model, X_train, Y_train, X_val, Y_val, epochs=2, batchSize=10, outfilepath=outfilepath, resfilepath=resfilepath)

In [34]:
# saveModel(model, outfilepath=outfilepath)

In [35]:
model.load_weights(f'{outfilepath}/modelW.weights.h5')

In [36]:
evaluateModel(model, X_test, Y_test, batchSize=10, outfilepath=outfilepath, resfilepath=resfilepath, val=False)

I0000 00:00:1736009398.661094      89 service.cc:145] XLA service 0x79ebc80250c0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1736009398.661158      89 service.cc:153]   StreamExecutor device (0): Tesla P100-PCIE-16GB, Compute Capability 6.0


  1/662 ━━━━━━━━━━━━━━━━━━━━ 3:10:15 17s/step

I0000 00:00:1736009414.904457      89 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


662/662 ━━━━━━━━━━━━━━━━━━━━ 200s 277ms/step
Jacard Index : 0.8836049601620388
Dice Coefficient : 0.9057411749147815


# Performance Comparison
|**Model**|**Dice Coefficient**|**Jaccard Index**|
|---------|--------------------|-----------------|
|MultiResUNet|0.897474389      |0.87823319367    |
|Modified U-Net|0.8914032548   |0.872448485      |
|Our Model|0.9057411749        |0.88360496       |